Introduce CI for AWS - part 2

wainersm opened the pull request:
Continuing [part 1](https://github.com/confidential-containers/cloud-api-adaptor/pull/2274) in this PR I've got:

- more failing tests are disabled
- securing the AWS account by leveraging OIDC short-lived credentials. The job will effectively run if `AWS_IAM_ROLE_ARN` secret is configured in the repository and some magic done on AWS account (https://docs.github.com/en/actions/how-tos/secure-your-work/security-harden-deployments/oidc-in-aws). Contain two commits that are being reviewed in https://github.com/confidential-containers/cloud-api-adaptor/pull/2551 that adds support to use temporary credentials
- added a post step to run clean up code for the case where the e2e tests were interrupted or crashed and the native teardown routine didn't run. Long-term I'm considering to remove that in favor of some regular clean-up routines on AWS account
 
On merging this PR I plan to enable the nightly job and leave future improvements to afterwards. Intentionally I'm not enabling AWS CI on pull requests yet because 1) want to have nightly running and stable first; and 2) still need to prepare the code to run in parallel jobs

ldoktor commented:
Seems fine to me although I don't feel confident enough to ack the go parts. Do you have a link to a testing pipeline results?

wainersm commented:
Hi @stevenhorsman , can I have your review on this one too?

wainersm commented on .github/workflows/e2e_aws.yaml:
```diff
@@ -165,6 +165,9 @@ jobs:
           echo "::group::Create cluster with $CONTAINER_RUNTIME"
+          # Use less space as possible so the Git runner doesn't run out of
+          # disk space.
+          export CLUSTER_DISK_SIZE=15
```
Hi @stevenhorsman , I've a commit prepared that applies the #2515 changes to aws, I should send in part 3. But I will drop this commit that reduces the size of the disk...

wainersm commented:
Update PR to address @stevenhorsman 's comments.

I tested in https://github.com/wainersm/cc-cloud-api-adaptor/actions/runs/17869189271/job/50819805331 and the test-aws is marked failed although the "Run tests" step is marked still running. Github cannot get the raw job logs at this moment. I suspect something is wrong with Github....

wainersm commented:
Hi @bpradipt , could you please approve?

wainersm commented:
Hi @stevenhorsman ,

I rebased and added one more commit: https://github.com/confidential-containers/cloud-api-adaptor/pull/2552/commits/3fc4d75d9b273ed9b2de4313fa62a3b6e187fbb7

I was going to send that commit in a part-3 but without it this part-2 is unusable. Let me know if your ack stands (or not). 

Tested in (passed): https://github.com/wainersm/cc-cloud-api-adaptor/actions/runs/18015834371/job/51261666730

bpradipt commented on src/cloud-api-adaptor/test/e2e/README.md:
```diff
@@ -131,6 +131,7 @@ Use the properties on the table below for AWS:
 |aws_vpc_sg_id|AWS VPC Security Groups ID||
```
It would be good to mention that e2e tests uses the tag `caa-e2e-test`. Also requirements for OIDC setup.

wainersm replied:
I will send it in a part-3 PR, ok?

wainersm commented on .github/workflows/e2e_aws.yaml:
```diff
@@ -34,19 +34,21 @@
-permissions: {}
+permissions:
+  # Required by aws-actions/configure-aws-credentials
+  id-token: write
```
I will move the comment of position to have zizmor happy.

wainersm commented on .github/workflows/e2e_aws.yaml:
```diff
@@ -56,7 +58,7 @@
         id: check_secrets
         run: |
-         if [[ -n "${{ secrets.AWS_ACCESS_KEY_ID }}" && -n "${{ secrets.AWS_SECRET_ACCESS_KEY }}" ]]; then
+         if [[ -n "${{ secrets.AWS_IAM_ROLE_ARN }}" ]]; then
```
I don't think we have a risk of code injection because we are reading from secrets. I could export the secret as an environment variable and zizmor will be happy (probably) but then I could be introducing a real security problem.

@stevenhorsman , you have hardened many workflows out there, what do you think?

stevenhorsman replied:
I think we ignore this alert - the code injection in general we've not fixed a lot of and I think you are right that making an env from a secret sounds less secure, so we'd need to dig in more to understand what zizmor wants, or if it's just not being that smart.

wainersm replied:
zizmor is just being extra-cautious (correctly IMHO) as stated in https://docs.zizmor.sh/audits/#template-injection:
"""
When used with a "pedantic" or "auditor" [persona](https://docs.zizmor.sh/usage/#using-personas), this audit will flag all template expansions in code contexts, even ones that are likely safe.

This is because zizmor considers all template expansions in code contexts to be code smells, and attempting to selectively permit them is more error-prone than forbidding them in a blanket fashion.
"""

stevenhorsman commented on .github/workflows/e2e_run_all.yaml:
```diff
@@ -47,7 +45,11 @@
-permissions: {}
+permissions:
+  # Required by aws-actions/configure-aws-credentials
+  id-token: write
```
Same here - this should be in the `aws` job only

wainersm commented on .github/workflows/e2e_aws.yaml:
```diff
@@ -34,19 +34,21 @@
-permissions: {}
+permissions:
+  # Required by aws-actions/configure-aws-credentials
+  id-token: write
```
I tried to allow id-token only at test-aws level but it didn't work. Maybe because test-aws is callable? Anyway, I will re-evaluate that as I might have made some mistake.

stevenhorsman replied:
Yeah, as the permission is on the job step I think that's the best practice?

stevenhorsman commented on .github/workflows/e2e_aws.yaml:
```diff
@@ -94,16 +100,19 @@ jobs:
         run: |
-           qcow2=$(echo "${{ inputs.podvm_image }}" | sed -e "s#.*/\(.*\):.*#\1.qcow2#")
-           ./hack/download-image.sh "${{ inputs.podvm_image }}" . -o "${qcow2}"
+           qcow2=$(echo "${PODVM_IMAGE}" | sed -e "s#.*/\(.*\):.*#\1.qcow2#")
```
I think this the line that shell check is failing on with: https://www.shellcheck.net/wiki/SC2001. It is interesting that it didn't complain previously though.

Don't ask me how look it took to work this out, but I think `qcow2=$(tagless=${PODVM_IMAGE%%:*}; echo ${tagless##*/})` might to what you are trying to do?
